### PR TITLE
Connect to and work with the prover service

### PR DESCRIPTION
**Pull request by iamrecursion**
This commit does some initial integration with the batch prover by connecting to `semaphore-mtb` and requesting proofs from it. It does not yet use these proofs in the batch prover.

### CODE REVIEW

File: src/contracts/batching/mod.rs [hunk 1]
```diff
@@ -57,27 +54,31 @@ impl IdentityManager for Contract {
...
-        todo!()
+        self.tree_depth
     }
 
     fn initial_leaf_value(&self) -> Field {
-        todo!()
+        self.initial_leaf_value
     }
 
     fn group_id(&self) -> U256 {
```
Comment (kustosz) [inline src/contracts/batching/mod.rs:78]: Just remove this altogether

File: src/prover/mod.rs [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+    /// # Arguments
+    /// - `start_index`: The index in the merkle tree at which the insertions
+    ///   were started.
+    /// - `pre_root`: The value of the merkle tree's root before identities were
+    ///   inserted.
+    /// - `post_root`: The value of the merkle tree's root after the identities
+    ///   were inserted.
+    /// - `identities`: A list of identity insertions, ordered in the order the
+    ///   identities were inserted into the merkle tree.
+    pub async fn generate_proof(
```
Comment (kustosz) [inline src/prover/mod.rs:71]: This should probably deal with batch sizes too?

File: src/prover/mod.rs [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+    /// The number of seconds to wait before timing out the transaction.
+    #[clap(long, env, default_value = "30")]
+    pub mtb_prover_timeout_secs: u64,
+}
+
+/// A representation of the connection to the MTB prover service.
+#[derive(Clone, Debug)]
+pub struct Prover {
+    target_url: Url,
+    client:     Client,
```
Comment (kustosz) [inline src/prover/mod.rs:39]: please use `reqwest::Client` qualified, `Client` means nothing

File: src/app.rs [hunk 4]
```diff
@@ -59,11 +63,14 @@ pub struct Options {
     pub ethereum: ethereum::Options,
 
     #[clap(flatten)]
-    pub contracts: contracts::Options,
+    pub identity_manager: contracts::Options,
```
Comment (kustosz) [inline src/app.rs:66]: I'm not sure about this rename. From the perspective of a deployer, this deals with the contract. Altogether, "identity manager" is a nebulous term that means much less than "contract" but what it actually means to mean is "contract".

**Review by kustosz — APPROVED** [review]
Good stuff, some comments inline

File: src/app.rs (hunk 4 above)
Comment (kustosz) [reply src/app.rs:66]: Also it was rolled back by Greg because it broke existing deployments

File: src/contracts/batching/mod.rs (hunk 1 above)
Comment (iamrecursion) [reply src/contracts/batching/mod.rs:78]: It'll be removed in the next PR where I pull out the legacy contract entirely.

**Pull request merged**